File: kits/dart/lib/src/harden.dart
import 'package:mcp_dart/mcp_dart.dart';

import 'messages.dart';
import 'strict_args.dart';

/// The mcp_dart adapter. The two operations that make "advertisement == runtime"
/// true are pure statics ([stamp] advertises the schema closed; [refusalFor]
/// enforces it at call time), so both are unit-testable without a live transport.
/// [HardenedServer] just wires them into an [McpServer].
class Harden {
  Harden._();

  /// Advertise closed: a copy of [schema] with `additionalProperties: false`
  /// when it is enforceable (has `properties`, not opted open); otherwise
  /// [schema] unchanged (an opted-open or unintrospectable schema is never
  /// quietly closed). This is what `tools/list` will show.
  static ToolInputSchema? stamp(ToolInputSchema? schema) {
    if (schema == null) return null;
    final json = schema.toJson();
    final enforced =
        json.containsKey('properties') && json['additionalProperties'] != true;
    if (!enforced) return schema;
    return ToolInputSchema(
      properties: schema.properties,
      required: schema.required,
      additionalProperties: false,
    );
  }

  /// The refusal result for a call carrying an undeclared argument, or `null`
  /// when the call is clean (or the schema is unenforceable / opted open). The
  /// message names the offending keys and NEVER echoes an argument value; the
  /// structured payload mirrors the Python/TS/PHP kits so the shared corpus
  /// passes identically.
  static CallToolResult? refusalFor(
    String name,
    ToolInputSchema? schema,
    Map<String, dynamic>? args, {
    String reconnectHint = Messages.defaultReconnectHint,
  }) {
    final r = StrictArgs.check(
      schema?.toJson(),
      args,
      toolName: name,
      reconnectHint: reconnectHint,
    );
    if (r.ok) return null;
    return CallToolResult(
      content: [TextContent(text: r.message!)],
      isError: true,
      structuredContent: {
        'ironmcp': {
          'refused': true,
          'tool': name,
          'unknown': r.unknown,
          'accepted': r.accepted,
        },
      },
    );
  }
}

/// Wraps an mcp_dart [McpServer] so every tool registered through it is hardened:
/// its input schema is advertised closed and an undeclared argument is refused
/// at call time instead of silently dropped. Tool bodies are untouched. Pass
/// [inner] to your `StreamableMcpServer`/serve path; register tools on the
/// wrapper.
class HardenedServer {
  HardenedServer(this.inner, {this.reconnectHint = Messages.defaultReconnectHint});

  final McpServer inner;
  final String reconnectHint;

  RegisteredTool registerTool(
    String name, {
    String? title,
    String? description,
    ToolInputSchema? inputSchema,
    ToolOutputSchema? outputSchema,
    ToolAnnotations? annotations,
    Map<String, dynamic>? meta,
    required ToolFunction callback,
  }) {
    return inner.registerTool(
      name,
      title: title,
      description: description,
      inputSchema: Harden.stamp(inputSchema),
      outputSchema: outputSchema,
      annotations: annotations,
      meta: meta,
      callback: (args, extra) {
        final refusal =
            Harden.refusalFor(name, inputSchema, args, reconnectHint: reconnectHint);
        return refusal ?? callback(args, extra);
      },
    );
  }
}

/// Wrap [server] for hardened tool registration.
HardenedServer harden(McpServer server,
        {String reconnectHint = Messages.defaultReconnectHint}) =>
    HardenedServer(server, reconnectHint: reconnectHint);

File: kits/php/src/Harden.php
<?php

declare(strict_types=1);

namespace IronMcp;

use Mcp\Capability\Registry;
use Mcp\Capability\RegistryInterface;
use Mcp\Schema\Tool;
use Mcp\Server;
use Mcp\Server\Builder;

/**
 * The one-call entry: make an mcp/sdk server refuse unknown tool arguments and advertise exactly
 * what it enforces. It adds the StrictArgsHandler (the runtime refusal, in the ironmcp shape) and
 * stamps additionalProperties:false onto every registered tool schema (advertise == runtime; the
 * SDK's own Opis validation then also refuses, as defence in depth). The stamp runs AFTER build,
 * because tool schemas are reflection-generated during discovery/build.
 */
final class Harden
{
    /**
     * Wrap a Builder: inject the guard + stamp the built registry closed. Returns the built server.
     * Provides the registry itself so the guard and the stamp share one populated registry.
     */
    public static function server(Builder $builder, string $reconnectHint = Messages::DEFAULT_RECONNECT_HINT): Server
    {
        return self::serveHardened($builder, reconnectHint: $reconnectHint)->server;
    }

    /**
     * The one-call convenience serve entry. Hardens the builder exactly as {@see server()} does
     * (strict-args guard + every tool schema stamped closed) AND, when [allowedHosts] is given,
     * builds the DNS-rebinding {@see HostGuard} (invariant #4) and returns it alongside the server so
     * a caller of the CONVENIENCE path — not only the low-level HostGuard constructor — can enable
     * the guard. PHP serving is PHP-FPM/PSR-15-shaped, so the returned {@see HardenedApp::$hostGuard}
     * is enforced by the app's HTTP front controller before dispatch (see {@see HardenedApp}).
     *
     * @param iterable<string>|null $allowedHosts hosts the server legitimately answers as; when
     *                                            non-null a default-deny HostGuard is built and any
     *                                            other (rebinding) Host is refused. null opts out.
     */
    public static function serveHardened(
        Builder $builder,
        ?iterable $allowedHosts = null,
        string $reconnectHint = Messages::DEFAULT_RECONNECT_HINT,
    ): HardenedApp {
        $registry = new Registry();
        $builder->setRegistry($registry);
        $builder->addRequestHandler(new StrictArgsHandler($registry, $reconnectHint));
        $server = $builder->build();
        self::registry($registry);

        $hostGuard = $allowedHosts !== null ? new HostGuard($allowedHosts) : null;

        return new HardenedApp($server, $hostGuard);
    }

    /** Stamp additionalProperties:false onto every registered tool schema (post-build primitive). */
    public static function registry(RegistryInterface $registry): void
    {
        // getTools() advertises Tool objects (the tools/list payload); getTool($name) returns the
        // ToolReference that also carries the handler needed to re-register.
        foreach ($registry->getTools()->references as $tool) {
            $closed = StrictArgs::stampClosed($tool->inputSchema);
            if ($closed === $tool->inputSchema) {
                continue; // opted open or unintrospectable — leave it
            }
            $handler = $registry->getTool($tool->name)->handler;
            $stamped = new Tool(
                $tool->name,
                $tool->title,
                $closed,
                $tool->description,
                $tool->annotations,
                $tool->icons,
                $tool->meta,
                $tool->outputSchema,
            );
            $registry->unregisterTool($tool->name);
            $registry->registerTool($stamped, $handler);
        }
    }
}
